### DISCUSSION TITLE
Make TimeoutBody Unpin

### DISCUSSION THREAD

**Discussion by hurlebouc** [tower-rs/tower-http#481] [Ideas]
Hi,

Currently `TimeoutBody` depends on `tokio::time::Sleep` which is `!Unpin`. This result is `TimeoutBody` not `Unpin` either.

This situation may be problematic, for instance, when using a `TimeoutBody` as an `AsyncBufRead` (thanks to `tokio_util::io::stream_reader::StreamReader` for instance) since majority of functions (like `read_line` for instance) ask for an `Unpin` stream.

It is not difficulte to make `StreamReader` Unpin by replacing `Sleep` by `Pin<Box<Sleep>>`. What do you thing about this change?

**Comment by hurlebouc** [comment]
Here is what could be changed: https://github.com/hurlebouc/tower-http/commit/7b69d2a8ef90a1b3ead018b2b1dfb8bb091dfddb

**Comment by nemosupremo** [comment]
I recently hit something like this, and eventually realized you could just call `Box::pin` on the body itself before passing it anything else. ex:

```rust
Box::pin(TimeoutBody::new(
    Duration::from_secs(30),
    req.into_body(),
)))
```
